src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -94,17 +102,50 @@ private static void ensureNoInputConflicts(CartesianProductTest annotation,
 		return sets;
 	}
 
-	private List<List<?>> getSetsFromRepeatableAnnotation(List<CartesianValueSource> valueSources) {
+	private List<List<?>> getSetsFromArgumentsSources(List<? extends Annotation> argumentsSources,
+			ExtensionContext context) {
 		List<List<?>> sets = new ArrayList<>();
-		for (CartesianValueSource source : valueSources) {
-			CartesianValueArgumentsProvider provider = new CartesianValueArgumentsProvider();
-			provider.accept(source);
-			List<Object> collect = provider.provideArguments().distinct().collect(toList());
-			sets.add(collect);
+		for (Annotation source : argumentsSources) {
+			sets.add(getSetFromAnnotation(context, source));
 		}
 		return sets;
 	}
 
+	private List<Object> getSetFromAnnotation(ExtensionContext context, Annotation source) {
```

Comment:
Is `Annotation source` a `...Rangesource`? If so, can you rename the variable(s) accordingly?

Reply:
Not a `RangeSource` - it is an `Annotation` that is itself annotated with `@ArgumentsSource`.

It works like this (basically copying `ParameterizedTest`, btw):
 - Your test is a `CartesianProductTest`
 - This test has annotations on it that are annotated with `@ArgumentsSource`
 - The `@ArgumentsSource` defines and `ArgumentsProvider` (class) that must implement `AnnotationConsumer<A extends Annotation>` (in our case this should be `CartesianAnnotationConsumer`)
 - We instantiate the `ArgumentsProvider` (must have no-args constructor)
 - We call the `accept(Annotation)` method with the _original_ annotation
 - We call `provideArguments(ExtensionContext context)`

Easy-peasy, right?

Reply:
In case this is not clear, here is an example:
```java
@ArgumentsSource(ExampleArgumentsProvider.class)
@interface ExampleSourceAnnotation {
   int intValue();
}

class ExampleArgumentsProvider implements ArgumentsProvider, CartesianAnnotationConsumer<ExampleSourceAnnotation> {
  int i;

  void accept(ExampleSourceAnnotation annotation) {
    /* do something with intValue() */
    i = annotation.intValue();
  }

  Stream<? extends Arguments> provideArguments(ExtensionContext context) {
    /* `accept` was already called so `i` should have a value */
    return Stream.of(Arguments.of(i));
  }
}
```

So, if we have a `CartesianProductTest`, annotated with `ExampleSourceAnnotation`, we store the annotation, instantiate `ExampleArgumentsProvider` (with no-arg constructor), 'feed it' the `ExampleSourceAnnotation` and call `provideArguments` to get the set for our test.